src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs

```diff
@@ -0,0 +1,290 @@
+            : base(threadingContext, assertIsForeground: false)
+        {
+            _threadingContext = threadingContext;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _asyncTasks = new JoinableTaskCollection(threadingContext.JoinableTaskContext);
+            _asyncTasks.Add(threadingContext.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var monitorSelectionService = (IVsMonitorSelection)await asyncServiceProvider.GetServiceAsync(typeof(SVsShellMonitorSelection));
+
+                if (ErrorHandler.Succeeded(monitorSelectionService.GetCurrentElementValue((uint)VSConstants.VSSELELEMID.SEID_DocumentFrame, out var value)))
+                {
+                    if (value is IVsWindowFrame windowFrame)
+                    {
+                        TrackWindowFrame(windowFrame);
+                        _activeFrame = windowFrame;
+                    }
+                }
+
+                monitorSelectionService.AdviseSelectionEvents(this, out var _);
+            }));
+        }
+
+        public void Dispose()
+        {
+            _asyncTasks.Join();
+        }
+        
+        /// <summary>
+        /// Raised when the set of window frames being tracked changes, which means the results from <see cref="GetActiveDocument"/> or <see cref="GetVisibleDocuments"/> may change.
+        /// </summary>
+        public event EventHandler DocumentsChanged;
+
+        /// <summary>
+        /// Returns the <see cref="DocumentId"/> of the active document in a given <see cref="Workspace"/>.
+        /// </summary>
+        public DocumentId GetActiveDocument(Workspace workspace)
+        {
+            ThisCanBeCalledOnAnyThread();
+
+            var visibleFramesSnapshot = _visibleFrames;
+            if (_activeFrame == null || visibleFramesSnapshot.IsEmpty)
+            {
+                return null;
+            }
+
+            foreach (var listener in visibleFramesSnapshot)
+            {
+                if (listener.Frame == _activeFrame)
+                {
+                    return listener.GetDocumentId(workspace);
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Get a read-only collection of the <see cref="DocumentId"/>s of all the visible documents in the given <see cref="Workspace"/>.
+        /// </summary>
+        public ImmutableArray<DocumentId> GetVisibleDocuments(Workspace workspace)
+        {
+            ThisCanBeCalledOnAnyThread();
+
+            var visibleFramesSnapshot = _visibleFrames;
+
+            var ids = ArrayBuilder<DocumentId>.GetInstance(visibleFramesSnapshot.Count);
+
+            foreach (var frame in visibleFramesSnapshot)
+            {
+                var documentId = frame.GetDocumentId(workspace);
```

Comment:
i assme we're using the solution snapshot in this workspace?  if so, this worries me as that coudl potentially change over time.  can we extract out the immutable part of the workspace we need up front and work withthat?

Reply:
@CyrusNajmabadi Not sure what you mean. Yes, this API returns just the ID, and it's up to the caller to line that back up to a Solution snapshot or deal with the unsynchronized nature. There is a set of extensions in IDocumentTrackingServiceExtensions that take the solution snapshot and call GetDocument to line it back up. Some callers use that, others don't if they are working with just IDs (like the solution crawler.)

Reply:
My main oogyness here was that literally the workspace could be changign as we're enumerating this collection.  Leading to getting documentIds that themselves refer may not even be consistent among themselves.  That feels supe r oogy to me.  But i agree it's not major as the caller has to decide what to do with all of them.

Reply:
It's written so any mismatch should just return null. This is an improvement over before, where it might return stuff from different workspaces entirely.

Reply:
Good point :D